ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/Snapshots.hs

```diff
@@ -180,32 +192,51 @@ diskSnapshotIsPermanent = isJust . dsSuffix
 diskSnapshotIsTemporary :: DiskSnapshot -> Bool
 diskSnapshotIsTemporary = not . diskSnapshotIsPermanent
 
--- | Read snapshot from disk
+data ReadSnaphotErr =
+    -- | Error while de-serialising data
+    ReadSnaphotFailed ReadIncrementalErr
+
+    -- | Checksum of read snapshot differs from the one tracked by @'DiskSnapshot'@
+  | ReadSnaphotBadCRC
+  deriving (Eq, Show)
+
+-- | Read snapshot from disk.
+--
+--   Fail on data corruption, i.e. when the checksum of the read data differs
+--   from the one tracked by @'DiskSnapshot'@, if any.
 readSnapshot ::
      forall m blk. IOLike m
   => SomeHasFS m
   -> (forall s. Decoder s (ExtLedgerState blk))
   -> (forall s. Decoder s (HeaderHash blk))
   -> DiskSnapshot
-  -> ExceptT ReadIncrementalErr m (ExtLedgerState blk)
-readSnapshot hasFS decLedger decHash =
-      ExceptT
-    . readIncremental hasFS decoder
-    . snapshotToPath
+  -> ExceptT ReadSnaphotErr m (ExtLedgerState blk)
+readSnapshot hasFS decLedger decHash snapshot = ExceptT $
+    readIncremental hasFS decoder (snapshotToPath snapshot) >>= \case
+       Left e  -> pure $ Left (ReadSnaphotFailed e)
+       Right (ledgerState, checksumAsRead) -> case dsChecksum snapshot of
+         Just checksumAsWritten ->
+           if checksumAsWritten /= checksumAsRead
+           then pure $ Left ReadSnaphotBadCRC
+           else pure (Right ledgerState)
+         Nothing                -> pure (Right ledgerState)
   where
     decoder :: Decoder s (ExtLedgerState blk)
     decoder = decodeSnapshotBackwardsCompatible (Proxy @blk) decLedger decHash
 
--- | Write snapshot to disk
+-- | Write a snapshot to disk, returning a @'DiskSnapshot'@ with the checksum of ledger state.
+--
+--   The checksum of the input @'DiskSnapshot'@, if any, is overridden.
 writeSnapshot ::
      forall m blk. MonadThrow m
   => SomeHasFS m
   -> (ExtLedgerState blk -> Encoding)
   -> DiskSnapshot
-  -> ExtLedgerState blk -> m ()
+  -> ExtLedgerState blk -> m DiskSnapshot
 writeSnapshot (SomeHasFS hasFS) encLedger ss cs = do
-    withFile hasFS (snapshotToPath ss) (WriteMode MustBeNew) $ \h ->
-      void $ hPut hasFS h $ CBOR.toBuilder (encode cs)
+    withFile hasFS (snapshotToPath ss) (WriteMode MustBeNew) $ \h -> do
+      (_, crc) <- hPutAllCRC hasFS h $ CBOR.toLazyByteString (encode cs)
+      pure ss{dsChecksum = Just crc}
```

Comment:
This can produce a CRC as a return type instead (the rest of the DiskSnapshot is actually an input parameter already). And I would argue that we don't want to return this but instead write it to a file.